streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -32,102 +33,198 @@
-        initialLoadedCheckpoints = checkpointFile.read();
-
-        log.trace("Checkpointable offsets read from checkpoint: {}", initialLoadedCheckpoints);
-
-        if (eosEnabled) {
-            // with EOS enabled, there should never be a checkpoint file _during_ processing.
-            // delete the checkpoint file after loading its stored offsets.
-            checkpointFile.delete();
-            checkpointFile = null;
-        }
+        this.baseDir = stateDirectory.directoryForTask(taskId);
+        this.checkpointFile = new OffsetCheckpoint(new File(baseDir, CHECKPOINT_FILE_NAME));
 
         log.debug("Created state store manager for task {}", taskId);
     }
 
+    void registerGlobalStateStores(final List<StateStore> stateStores) {
+        log.debug("Register global stores {}", stateStores);
+        for (final StateStore stateStore : stateStores) {
+            globalStores.put(stateStore.name(), stateStore);
+        }
+    }
+
+    @Override
+    public StateStore getGlobalStore(final String name) {
+        return globalStores.get(name);
+    }
 
-    public static String storeChangelogTopic(final String applicationId,
-                                             final String storeName) {
-        return applicationId + "-" + storeName + STATE_CHANGELOG_TOPIC_SUFFIX;
+    // package-private for test only
+    void initializeStoreOffsetsFromCheckpoint() {
+        try {
+            final Map<TopicPartition, Long> loadedCheckpoints = checkpointFile.read();
+
+            log.trace("Loaded offsets from the checkpoint file: {}", loadedCheckpoints);
+
+            for (final StateStoreMetadata store : stores.values()) {
+                if (store.changelogPartition == null) {
+                    log.info("State store {} is not logged and hence would not be restored", store.stateStore.name());
+                } else {
+                    if (loadedCheckpoints.containsKey(store.changelogPartition)) {
+                        store.setOffset(loadedCheckpoints.remove(store.changelogPartition));
+
+                        log.debug("State store {} initialized from checkpoint with offset {} at changelog {}",
+                            store.stateStore.name(), store.offset, store.changelogPartition);
+                    } else {
+                        // TODO K9113: for EOS when there's no checkpointed offset, we should treat it as TaskCorrupted
+
+                        log.info("State store {} did not find checkpoint offset, hence would " +
+                                "default to the starting offset at changelog {}",
+                            store.stateStore.name(), store.changelogPartition);
+                    }
+                }
+            }
+
+            if (!loadedCheckpoints.isEmpty()) {
+                log.warn("Some loaded checkpoint offsets cannot find their corresponding state stores: {}", loadedCheckpoints);
+            }
+
+            checkpointFile.delete();
```

Comment:
in retrospect, I like the idea of doing this regardless of EOS. Why should we deliberately produce wrong results in ALO mode? We can certainly optimize to be able to use semi-trustworthy data, but let's treat that separately from EOS vs ALO

Reply:
The idea is that for ALO if we failed to write the first checkpoint after restarting we can still fallback to the original checkpoint even though the store may have been updated and hence there would be duplicates. But since the window gap (just one commit interval) is so small I think it does not worth making the code more complicated with EO v.s. ALO.